File: src/ai/neural/encoding.ts
/**
 * Position Encoding for Neural Network Input
 *
 * Provides different encoding schemes for representing Connect Four positions
 * as neural network inputs.
 */

import { type Board, ROWS, COLUMNS, type Player, type GameState } from '../../game/makefour'
import type { EncodedPosition, PositionEncodingType } from './interface'

/**
 * Encodes a position using one-hot encoding with 3 channels.
 * Shape: [1, 3, 6, 7] = [batch, channels, rows, columns]
 *
 * Channel 0: Player 1's pieces (1 where P1 has a piece, 0 elsewhere)
 * Channel 1: Player 2's pieces (1 where P2 has a piece, 0 elsewhere)
 * Channel 2: Current player indicator (all 1s if P1 to move, all 0s if P2)
 *
 * This encoding preserves spatial relationships for CNNs.
 */
export function encodeOneHot(state: GameState): EncodedPosition {
  const { board, currentPlayer, moveHistory } = state
  const channelSize = ROWS * COLUMNS // 42
  const totalSize = 3 * channelSize // 126

  const input = new Float32Array(totalSize)

  // Channel 0: Player 1 pieces
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      if (board[row][col] === 1) {
        input[row * COLUMNS + col] = 1
      }
    }
  }

  // Channel 1: Player 2 pieces
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      if (board[row][col] === 2) {
        input[channelSize + row * COLUMNS + col] = 1
      }
    }
  }

  // Channel 2: Current player (all 1s if player 1 to move)
  if (currentPlayer === 1) {
    for (let i = 0; i < channelSize; i++) {
      input[2 * channelSize + i] = 1
    }
  }

  return {
    input,
    shape: [1, 3, ROWS, COLUMNS],
    currentPlayer,
    moveCount: moveHistory.length,
  }
}

/**
 * Encodes a position using bitboard representation.
 * Shape: [1, 2] where each element is a 49-bit integer represented as two Float32s.
 *
 * Each player's position is represented as a bitmask of occupied cells.
 * Cell (row, col) maps to bit (row * 7 + col).
 */
export function encodeBitboard(state: GameState): EncodedPosition {
  const { board, currentPlayer, moveHistory } = state

  // Use two numbers to represent each 42-bit bitboard
  // (JavaScript numbers are 64-bit floats, safe integers up to 2^53)
  let p1Bits = 0
  let p2Bits = 0

  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      const bit = row * COLUMNS + col
      if (board[row][col] === 1) {
        p1Bits |= 1 << bit
      } else if (board[row][col] === 2) {
        p2Bits |= 1 << bit
      }
    }
  }

  // Include current player info
  const playerBit = currentPlayer === 1 ? 1 : 0

  const input = new Float32Array([p1Bits, p2Bits, playerBit, moveHistory.length])

  return {
    input,
    shape: [1, 4],
    currentPlayer,
    moveCount: moveHistory.length,
  }
}

/**
 * Encodes a position as a flat binary vector.
 * Shape: [1, 85] = [batch, features]
 *
 * Features:
 * - 42 bits for player 1 pieces
 * - 42 bits for player 2 pieces
 * - 1 bit for current player
 *
 * Simple and efficient for MLP architectures.
 */
export function encodeFlatBinary(state: GameState): EncodedPosition {
  const { board, currentPlayer, moveHistory } = state
  const cellCount = ROWS * COLUMNS // 42
  const totalSize = cellCount * 2 + 1 // 85

  const input = new Float32Array(totalSize)

  // First 42: Player 1 pieces
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      if (board[row][col] === 1) {
        input[row * COLUMNS + col] = 1
      }
    }
  }

  // Next 42: Player 2 pieces
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      if (board[row][col] === 2) {
        input[cellCount + row * COLUMNS + col] = 1
      }
    }
  }

  // Last bit: Current player (1 if player 1, 0 if player 2)
  input[cellCount * 2] = currentPlayer === 1 ? 1 : 0

  return {
    input,
    shape: [1, totalSize],
    currentPlayer,
    moveCount: moveHistory.length,
  }
}

/**
 * Encodes a position using the specified encoding type.
 */
export function encodePosition(
  state: GameState,
  encoding: PositionEncodingType
): EncodedPosition {
  switch (encoding) {
    case 'onehot-6x7x3':
      return encodeOneHot(state)
    case 'bitboard':
      return encodeBitboard(state)
    case 'flat-binary':
      return encodeFlatBinary(state)
    default:
      throw new Error(`Unknown encoding type: ${encoding}`)
  }
}

/**
 * Decodes a one-hot encoded position back to a board state.
 * Useful for debugging and visualization.
 */
export function decodeOneHot(encoded: EncodedPosition): { board: Board; currentPlayer: Player } {
  if (encoded.shape[1] !== 3 || encoded.shape[2] !== ROWS || encoded.shape[3] !== COLUMNS) {
    throw new Error('Invalid one-hot encoding shape')
  }

  const channelSize = ROWS * COLUMNS
  const board: Board = Array.from({ length: ROWS }, () => Array(COLUMNS).fill(null))

  // Decode player 1 pieces
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      if (encoded.input[row * COLUMNS + col] === 1) {
        board[row][col] = 1
      }
    }
  }

  // Decode player 2 pieces
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLUMNS; col++) {
      if (encoded.input[channelSize + row * COLUMNS + col] === 1) {
        board[row][col] = 2
      }
    }
  }

  // Decode current player
  const currentPlayer: Player = encoded.input[2 * channelSize] === 1 ? 1 : 2

  return { board, currentPlayer }
}

/**
 * Converts neural network output probabilities to valid move probabilities.
 * Masks out invalid moves (full columns) and renormalizes.
 *
 * @param rawPolicy - Raw policy output from neural network (7 values)
 * @param board - Current board state
 * @returns Normalized probabilities for valid moves only
 */
export function maskInvalidMoves(rawPolicy: number[], board: Board): number[] {
  const masked = rawPolicy.map((prob, col) => {
    // Check if column is full (top row occupied)
    if (board[0][col] !== null) {
      return 0
    }
    return Math.max(0, prob) // Ensure non-negative
  })

  // Renormalize
  const sum = masked.reduce((a, b) => a + b, 0)
  if (sum === 0) {
    // No valid moves - shouldn't happen in normal gameplay
    return masked
  }

  return masked.map((p) => p / sum)
}

/**
 * Selects a move from policy probabilities.
 *
 * @param policy - Move probabilities (7 values)
 * @param temperature - Temperature for sampling (0 = deterministic, higher = more random)
 * @returns Selected column index (0-6)
 */
export function sampleFromPolicy(policy: number[], temperature = 0): number {
  if (temperature === 0) {
    // Deterministic: pick highest probability
    let maxProb = -1
    let maxCol = 0
    for (let col = 0; col < policy.length; col++) {
      if (policy[col] > maxProb) {
        maxProb = policy[col]
        maxCol = col
      }
    }
    return maxCol
  }

  // Temperature-scaled sampling
  const scaled = policy.map((p) => p ** (1 / temperature))
  const sum = scaled.reduce((a, b) => a + b, 0)
  const normalized = scaled.map((p) => p / sum)

  // Random sampling
  const rand = Math.random()
  let cumulative = 0
  for (let col = 0; col < normalized.length; col++) {
    cumulative += normalized[col]
    if (rand < cumulative) {
      return col
    }
  }

  return normalized.length - 1
}

/**
 * Gets the input shape for a given encoding type.
 */
export function getInputShape(encoding: PositionEncodingType): number[] {
  switch (encoding) {
    case 'onehot-6x7x3':
      return [1, 3, ROWS, COLUMNS]
    case 'bitboard':
      return [1, 4]
    case 'flat-binary':
      return [1, 85]
    default:
      throw new Error(`Unknown encoding type: ${encoding}`)
  }
}

/**
 * Gets the total number of input features for a given encoding type.
 */
export function getInputSize(encoding: PositionEncodingType): number {
  const shape = getInputShape(encoding)
  return shape.reduce((a, b) => a * b, 1)
}

File: training/src/data/encoding.py
"""
Position Encoding for Neural Network Input

Provides different encoding schemes for representing Connect Four positions
as neural network inputs. These MUST match the TypeScript encodings in
src/ai/neural/encoding.ts exactly.
"""

from typing import Literal, TypeAlias
import numpy as np

# Board dimensions (must match TypeScript)
ROWS = 6
COLUMNS = 7

# Type aliases
Board: TypeAlias = list[list[int | None]]  # 6x7, None=empty, 1=player1, 2=player2
Player: TypeAlias = Literal[1, 2]
EncodingType: TypeAlias = Literal["onehot-6x7x3", "flat-binary", "bitboard"]


def encode_onehot(board: Board, current_player: Player) -> np.ndarray:
    """
    Encodes a position using one-hot encoding with 3 channels.
    Shape: [3, 6, 7] = [channels, rows, columns]

    Channel 0: Player 1's pieces (1 where P1 has a piece, 0 elsewhere)
    Channel 1: Player 2's pieces (1 where P2 has a piece, 0 elsewhere)
    Channel 2: Current player indicator (all 1s if P1 to move, all 0s if P2)

    This encoding preserves spatial relationships for CNNs.

    IMPORTANT: This must match encodeOneHot() in src/ai/neural/encoding.ts
    """
    channel_size = ROWS * COLUMNS  # 42
    total_size = 3 * channel_size  # 126

    # Create flat array (matches TypeScript Float32Array layout)
    encoded = np.zeros(total_size, dtype=np.float32)

    # Channel 0: Player 1 pieces
    for row in range(ROWS):
        for col in range(COLUMNS):
            if board[row][col] == 1:
                encoded[row * COLUMNS + col] = 1.0

    # Channel 1: Player 2 pieces
    for row in range(ROWS):
        for col in range(COLUMNS):
            if board[row][col] == 2:
                encoded[channel_size + row * COLUMNS + col] = 1.0

    # Channel 2: Current player (all 1s if player 1 to move)
    if current_player == 1:
        for i in range(channel_size):
            encoded[2 * channel_size + i] = 1.0

    return encoded


def encode_onehot_3d(board: Board, current_player: Player) -> np.ndarray:
    """
    Same as encode_onehot but returns shape [3, 6, 7] for CNN input.
    """
    flat = encode_onehot(board, current_player)
    return flat.reshape(3, ROWS, COLUMNS)


def encode_flat_binary(board: Board, current_player: Player) -> np.ndarray:
    """
    Encodes a position as a flat binary vector.
    Shape: [85] = [features]

    Features:
    - 42 bits for player 1 pieces
    - 42 bits for player 2 pieces
    - 1 bit for current player

    Simple and efficient for MLP architectures.

    IMPORTANT: This must match encodeFlatBinary() in src/ai/neural/encoding.ts
    """
    cell_count = ROWS * COLUMNS  # 42
    total_size = cell_count * 2 + 1  # 85

    encoded = np.zeros(total_size, dtype=np.float32)

    # First 42: Player 1 pieces
    for row in range(ROWS):
        for col in range(COLUMNS):
            if board[row][col] == 1:
                encoded[row * COLUMNS + col] = 1.0

    # Next 42: Player 2 pieces
    for row in range(ROWS):
        for col in range(COLUMNS):
            if board[row][col] == 2:
                encoded[cell_count + row * COLUMNS + col] = 1.0

    # Last bit: Current player (1 if player 1, 0 if player 2)
    encoded[cell_count * 2] = 1.0 if current_player == 1 else 0.0

    return encoded


def encode_bitboard(board: Board, current_player: Player, move_count: int = 0) -> np.ndarray:
    """
    Encodes a position using bitboard representation.
    Shape: [4] where elements are [p1_bits, p2_bits, player_bit, move_count]

    Each player's position is represented as a bitmask of occupied cells.
    Cell (row, col) maps to bit (row * 7 + col).

    IMPORTANT: This must match encodeBitboard() in src/ai/neural/encoding.ts
    """
    p1_bits = 0
    p2_bits = 0

    for row in range(ROWS):
        for col in range(COLUMNS):
            bit = row * COLUMNS + col
            if board[row][col] == 1:
                p1_bits |= 1 << bit
            elif board[row][col] == 2:
                p2_bits |= 1 << bit

    # Include current player info
    player_bit = 1 if current_player == 1 else 0

    return np.array([p1_bits, p2_bits, player_bit, move_count], dtype=np.float32)


def decode_onehot(encoded: np.ndarray) -> tuple[Board, Player]:
    """
    Decodes a one-hot encoded position back to a board state.
    Useful for debugging and visualization.

    IMPORTANT: This must match decodeOneHot() in src/ai/neural/encoding.ts
    """
    # Handle both flat [126] and shaped [3, 6, 7] input
    if encoded.shape == (3, ROWS, COLUMNS):
        flat = encoded.flatten()
    elif encoded.shape == (126,):
        flat = encoded
    else:
        raise ValueError(f"Invalid one-hot encoding shape: {encoded.shape}")

    channel_size = ROWS * COLUMNS
    board: Board = [[None for _ in range(COLUMNS)] for _ in range(ROWS)]

    # Decode player 1 pieces
    for row in range(ROWS):
        for col in range(COLUMNS):
            if flat[row * COLUMNS + col] == 1:
                board[row][col] = 1

    # Decode player 2 pieces
    for row in range(ROWS):
        for col in range(COLUMNS):
            if flat[channel_size + row * COLUMNS + col] == 1:
                board[row][col] = 2

    # Decode current player
    current_player: Player = 1 if flat[2 * channel_size] == 1 else 2

    return board, current_player


def encode_position(
    board: Board,
    current_player: Player,
    encoding: EncodingType,
    move_count: int = 0,
) -> np.ndarray:
    """
    Encodes a position using the specified encoding type.
    """
    if encoding == "onehot-6x7x3":
        return encode_onehot(board, current_player)
    elif encoding == "flat-binary":
        return encode_flat_binary(board, current_player)
    elif encoding == "bitboard":
        return encode_bitboard(board, current_player, move_count)
    else:
        raise ValueError(f"Unknown encoding type: {encoding}")


def get_input_shape(encoding: EncodingType) -> tuple[int, ...]:
    """
    Gets the input shape for a given encoding type.
    """
    if encoding == "onehot-6x7x3":
        return (1, 3, ROWS, COLUMNS)
    elif encoding == "flat-binary":
        return (1, 85)
    elif encoding == "bitboard":
        return (1, 4)
    else:
        raise ValueError(f"Unknown encoding type: {encoding}")


def get_input_size(encoding: EncodingType) -> int:
    """
    Gets the total number of input features for a given encoding type.
    """
    shape = get_input_shape(encoding)
    result = 1
    for dim in shape:
        result *= dim
    return result


def board_from_moves(moves: list[int]) -> tuple[Board, Player]:
    """
    Reconstructs a board state from a sequence of moves.

    Args:
        moves: List of column indices (0-6) representing each move in order.

    Returns:
        Tuple of (board, current_player) after all moves are applied.
    """
    board: Board = [[None for _ in range(COLUMNS)] for _ in range(ROWS)]
    current_player: Player = 1

    for col in moves:
        # Find the lowest empty row in the column
        for row in range(ROWS - 1, -1, -1):
            if board[row][col] is None:
                board[row][col] = current_player
                break
        # Alternate players
        current_player = 2 if current_player == 1 else 1

    return board, current_player


def board_to_string(board: Board) -> str:
    """
    Converts a board to a human-readable string representation.
    """
    symbols = {None: ".", 1: "X", 2: "O"}
    lines = []
    for row in board:
        lines.append(" ".join(symbols[cell] for cell in row))
    lines.append("0 1 2 3 4 5 6")
    return "\n".join(lines)


def flip_board_horizontal(board: Board) -> Board:
    """
    Flips a board horizontally (mirrors columns).
    This is a valid data augmentation for Connect Four.
    """
    return [list(reversed(row)) for row in board]


def flip_move_horizontal(move: int) -> int:
    """
    Flips a move horizontally (mirrors column index).
    """
    return COLUMNS - 1 - move
